From 841787d44d88e2e60171c755876e833974b56d3e Mon Sep 17 00:00:00 2001
From: Italo Valcy <italo@ampath.net>
Date: Fri, 27 Aug 2021 11:17:57 -0300
Subject: [PATCH 1/5] Using Lock to avoid race conditions when saving the
 topology status

---
 main.py | 28 +++++++++++++++++-----------
 1 file changed, 17 insertions(+), 11 deletions(-)

diff --git a/main.py b/main.py
index 6f88e29..89ae8db 100644
--- a/main.py
+++ b/main.py
@@ -3,6 +3,7 @@
 Manage the network topology
 """
 import time
+from threading import Lock
 
 from flask import jsonify, request
 from werkzeug.exceptions import BadRequest, UnsupportedMediaType
@@ -48,9 +49,13 @@ def setup(self):
 
         self.storehouse = StoreHouse(self.controller)
 
+        self._lock = Lock()
+
+    @listen_to('kytos/storehouse.loaded')
     def execute(self):
         """Execute once when the napp is running."""
-        self._load_network_status()
+        with self._lock:
+            self._load_network_status()
 
     def shutdown(self):
         """Do nothing."""
@@ -708,16 +713,17 @@ def add_links(self, event):
     @listen_to('.*.network_status.updated')
     def save_status_on_storehouse(self, event=None):
         """Save the network administrative status using storehouse."""
-        status = self._get_switches_dict()
-        status['id'] = 'network_status'
-        if event:
-            content = event.content
-            log.info(f"Storing the administrative state of the"
-                     f" {content['attribute']} attribute to"
-                     f" {content['state']} in the interfaces"
-                     f" {content['interface_ids']}")
-        status.update(self._get_links_dict())
-        self.storehouse.save_status(status)
+        with self._lock:
+            status = self._get_switches_dict()
+            status['id'] = 'network_status'
+            if event:
+                content = event.content
+                log.info(f"Storing the administrative state of the"
+                         f" {content['attribute']} attribute to"
+                         f" {content['state']} in the interfaces"
+                         f" {content['interface_ids']}")
+            status.update(self._get_links_dict())
+            self.storehouse.save_status(status)
 
     def notify_switch_enabled(self, dpid):
         """Send an event to notify that a switch is enabled."""

From 49058619cf7d0358364e85e9a057a79731dcdc40 Mon Sep 17 00:00:00 2001
From: Italo Valcy <italo@ampath.net>
Date: Fri, 27 Aug 2021 11:19:04 -0300
Subject: [PATCH 2/5] refactoring the restore topology routing to load on Kytos
 startup

---
 main.py | 191 +++++++++++++++++++++++---------------------------------
 1 file changed, 77 insertions(+), 114 deletions(-)

diff --git a/main.py b/main.py
index 89ae8db..28b8a3a 100644
--- a/main.py
+++ b/main.py
@@ -20,8 +20,6 @@
 from napps.kytos.topology.storehouse import StoreHouse
 
 DEFAULT_LINK_UP_TIMER = 10
-DEFAULT_INTERFACE_RESTORE_TIMER = 2
-RESTORE_INTERFACE_ATTEMPTS = 20
 
 
 class Main(KytosNApp):  # pylint: disable=too-many-public-methods
@@ -34,14 +32,8 @@ def setup(self):
         """Initialize the NApp's links list."""
         self.links = {}
         self.store_items = {}
-        self.switches_state = {}
-        self.interfaces_state = {}
-        self.links_state = {}
-        self._verified_links = []
         self.link_up_timer = getattr(settings, 'LINK_UP_TIMER',
                                      DEFAULT_LINK_UP_TIMER)
-        self.interface_restore = getattr(settings, 'INTERFACE_RESTORE_TIMER',
-                                         DEFAULT_INTERFACE_RESTORE_TIMER)
 
         self.verify_storehouse('switches')
         self.verify_storehouse('interfaces')
@@ -126,94 +118,74 @@ def _get_link_from_interface(self, interface):
                 return link
         return None
 
-    def _restore_link(self, link_id):
-        """Restore link's administrative state from storehouse."""
-        try:
-            state = self.links_state[link_id]
-        except KeyError:
-            error = (f'The link {link_id} has no stored '
-                     'administrative state to be restored.')
-            raise RestoreError(error)
+    def _load_link(self, link_att):
+        dpid_a = link_att['endpoint_a']['switch']
+        dpid_b = link_att['endpoint_b']['switch']
+        port_a = link_att['endpoint_a']['port_number']
+        port_b = link_att['endpoint_b']['port_number']
+        link_str = f'{dpid_a}:{port_a}-{dpid_b}:{port_b}'
+        log.info(f'Loading link from storehouse {link_str}')
 
         try:
-            link = self.links[link_id]
-            if state['enabled']:
-                link.enable()
-            else:
-                link.disable()
-        except KeyError:
-            error = ('Error restoring link status.'
-                     f'The link {link_id} does not exist.')
+            switch_a = self.controller.switches[dpid_a]
+            switch_b = self.controller.switches[dpid_b]
+            interface_a = switch_a.interfaces[port_a]
+            interface_b = switch_b.interfaces[port_b]
+        except Exception as err:
+            error = f'Fail to load endpoints for link {link_str}: {err}'
             raise RestoreError(error)
-        log.info(f'The state of link {link.id} has been restored.')
-        self.notify_topology_update()
-        self.update_instance_metadata(link)
-        self.notify_link_status_change(link, reason='from storehouse')
 
-    def _restore_switch(self, switch_id):
-        """Restore switch's administrative state from storehouse."""
-        try:
-            state = self.switches_state[switch_id]
-        except KeyError:
-            error = (f'The switch {switch_id} has no stored'
-                     ' administrative state to be restored.')
-            raise RestoreError(error)
+        link = self._get_link_or_create(interface_a, interface_b)
 
-        try:
-            switch = self.controller.switches[switch_id]
-        except KeyError:
-            # Maybe we should remove the switch from switches_state here
-            error = ('Error while restoring switches status. The '
-                     f'switch {switch_id} does not exist.')
-            raise RestoreError(error)
+        if link_att['enabled']:
+            link.enable()
+        else:
+            link.disable()
 
-        if state:
+        interface_a.update_link(link)
+        interface_b.update_link(link)
+        interface_a.nni = True
+        interface_b.nni = True
+        self.update_instance_metadata(link)
+
+    def _load_switch(self, switch_id, switch_att):
+        log.info(f'Loading switch from storehouse dpid={switch_id}')
+        switch = self.controller.get_switch_or_create(switch_id)
+        if switch_att['enabled']:
             switch.enable()
-            self.notify_switch_enabled(switch_id)
         else:
             switch.disable()
-            self.notify_switch_disabled(switch_id)
-
-        log.debug('Waiting to restore administrative state of switch '
-                  f'{switch_id} interfaces.')
-        i = 0
-        # wait to restore interfaces
-        while not switch.interfaces and i < RESTORE_INTERFACE_ATTEMPTS:
-            time.sleep(self.interface_restore)
-            i += 1
-        if not switch.interfaces:
-            error = ('Error restoring administrative state of switch '
-                     f'{switch_id} interfaces.')
-            raise RestoreError(error)
-
-        # restore interfaces
-        for interface_id in switch.interfaces:
-            iface_id = ":".join([switch_id, str(interface_id)])
-            # restore only the administrative state of saved interfaces
-            if iface_id not in self.interfaces_state:
-                error = ("The stored topology is different from the current "
-                         f"topology. The interface {iface_id} hasn't been "
-                         "stored.")
-                log.info(error)
-                continue
-            state = self.interfaces_state[iface_id]
-            iface_number = int(interface_id)
-            iface_status, lldp_status = state
-            try:
-                interface = switch.interfaces[iface_number]
-            except KeyError:
-                log.error('Error restoring interface status: '
-                          '%s does not exist.', iface_id)
-                continue
+        switch.description['manufacturer'] = switch_att.get('manufacturer', '')
+        switch.description['hardware'] = switch_att.get('hardware', '')
+        switch.description['software'] = switch_att.get('software')
+        switch.description['serial'] = switch_att.get('serial', '')
+        switch.description['data_path'] = switch_att.get('data_path', '')
+        self.update_instance_metadata(switch)
 
-            if iface_status:
+        for iface_id, iface_att in switch_att.get('interfaces', {}).items():
+            log.info(f'Loading interface iface_id={iface_id}')
+            interface = switch.update_or_create_interface(
+                            port_no=iface_att['port_number'],
+                            name=iface_att['name'],
+                            address=iface_att.get('mac', None),
+                            speed=iface_att.get('speed', None))
+            if iface_att['enabled']:
                 interface.enable()
             else:
                 interface.disable()
-            interface.lldp = lldp_status
+            interface.lldp = iface_att['lldp']
             self.update_instance_metadata(interface)
-
-        log.info(f'The state of switch {switch_id} has been restored.')
+            name = 'kytos/topology.port.created'
+            event = KytosEvent(name=name, content={
+                                              'switch': switch_id,
+                                              'port': interface.port_number,
+                                              'port_description': {
+                                                  'alias': interface.name,
+                                                  'mac': interface.address,
+                                                  'state': interface.state
+                                                  }
+                                              })
+            self.controller.buffers.app.put(event)
 
     # pylint: disable=attribute-defined-outside-init
     def _load_network_status(self):
@@ -221,26 +193,31 @@ def _load_network_status(self):
         try:
             status = self.storehouse.get_data()
         except FileNotFoundError as error:
-            log.info(error)
+            log.error(f'Fail to load network status from storehouse: {error}')
             return
-        if status:
-            switches = status['network_status']['switches']
-            self.links_state = status['network_status']['links']
-
-            for switch_id, switch_att in switches.items():
-                # get switches status
-                self.switches_state[switch_id] = switch_att['enabled']
-                iface = switch_att['interfaces']
-                # get interface status
-                for iface_id, iface_att in iface.items():
-                    enabled_value = iface_att['enabled']
-                    lldp_value = iface_att['lldp']
-                    self.interfaces_state[iface_id] = (enabled_value,
-                                                       lldp_value)
 
-        else:
-            error = 'There is no status saved to restore.'
-            log.info(error)
+        if not status:
+            log.info('There is no status saved to restore.')
+            return
+
+        switches = status['network_status']['switches']
+        links = status['network_status']['links']
+
+        log.debug("_load_network_status switches=%s" % switches)
+        for switch_id, switch_att in switches.items():
+            try:
+                self._load_switch(switch_id, switch_att)
+            # pylint: disable=broad-except
+            except Exception as err:
+                log.error(f'Error loading switch: {err}')
+
+        log.debug("_load_network_status links=%s" % links)
+        for link_id, link_att in links.items():
+            try:
+                self._load_link(link_att)
+            # pylint: disable=broad-except
+            except Exception as err:
+                log.error(f'Error loading link {link_id}: {err}')
 
     @rest('v3/')
     def get_topology(self):
@@ -250,18 +227,6 @@ def get_topology(self):
         """
         return jsonify(self._get_topology_dict())
 
-    def restore_network_status(self, obj):
-        """Restore the network administrative status saved in storehouse."""
-        try:
-            if isinstance(obj, Switch):
-                self._restore_switch(obj.id)
-            elif isinstance(obj, Link):
-                if obj.id not in self._verified_links:
-                    self._verified_links.append(obj.id)
-                    self._restore_link(obj.id)
-        except RestoreError as exc:
-            log.debug(exc)
-
     # Switch related methods
     @rest('v3/switches')
     def get_switches(self):
@@ -546,7 +511,6 @@ def handle_new_switch(self, event):
         log.debug('Switch %s added to the Topology.', switch.id)
         self.notify_topology_update()
         self.update_instance_metadata(switch)
-        self.restore_network_status(switch)
         if switch.is_enabled():
             self.notify_switch_enabled(switch.id)
 
@@ -690,7 +654,6 @@ def add_links(self, event):
         interface_b.nni = True
 
         self.notify_topology_update()
-        self.restore_network_status(link)
 
     # def add_host(self, event):
     #    """Update the topology with a new Host."""

From 7e535f1c45812b04279bb4d154af857f68150d47 Mon Sep 17 00:00:00 2001
From: Italo Valcy <italo@ampath.net>
Date: Fri, 27 Aug 2021 11:19:21 -0300
Subject: [PATCH 3/5] clean settings

---
 settings.py | 2 --
 1 file changed, 2 deletions(-)

diff --git a/settings.py b/settings.py
index f756511..290b26d 100644
--- a/settings.py
+++ b/settings.py
@@ -7,5 +7,3 @@
 LINK_UP_TIMER = 10
 # Time (in seconds) to wait retrieve box from storehouse
 BOX_RESTORE_TIMER = 0.1
-# Time (in seconds) to wait before restore interfaces
-INTERFACE_RESTORE_TIMER = 2

From 8951b6d7d336093b8ca2beb6a6f088777985a606 Mon Sep 17 00:00:00 2001
From: Italo Valcy <italo@ampath.net>
Date: Fri, 27 Aug 2021 11:19:32 -0300
Subject: [PATCH 4/5] adjusting unit tests

---
 tests/unit/test_main.py | 405 +++++++++++++++++++++++++++++++---------
 1 file changed, 312 insertions(+), 93 deletions(-)

diff --git a/tests/unit/test_main.py b/tests/unit/test_main.py
index 482b777..92639a7 100644
--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -41,6 +41,7 @@ def test_get_event_listeners(self):
                            'kytos/maintenance.end_link',
                            'kytos/maintenance.start_switch',
                            'kytos/maintenance.end_switch',
+                           'kytos/storehouse.loaded',
                            '.*.network_status.updated',
                            '.*.interface.is.nni',
                            '.*.connection.lost',
@@ -190,138 +191,356 @@ def test_get_topology(self):
     @patch('napps.kytos.topology.main.StoreHouse.get_data')
     def test_load_network_status(self, mock_storehouse_get_data):
         """Test _load_network_status."""
+        link_id = \
+            'cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260'
+        dpid_a = '00:00:00:00:00:00:00:01'
+        dpid_b = '00:00:00:00:00:00:00:02'
         status = {
             'network_status': {
                 'id': 'network_status',
                 'links': {
-                    '4d42dc08522': {
+                    link_id: {
                         'enabled': True,
                         'endpoint_a': {
-                            'switch': '00:00:00:00:00:00:00:01',
-                            'id': '00:00:00:00:00:00:00:00:1'
+                            'switch': dpid_a,
+                            'port_number': 2
                         },
                         'endpoint_b': {
-                            'switch': '00:00:00:00:00:00:00:01',
-                            'id': '00:00:00:00:00:00:00:00:2'
+                            'switch': dpid_b,
+                            'port_number': 2
                         }
                     }
                 },
                 'switches': {
-                    '00:00:00:00:00:00:00:01': {
-                        'dpid': '00:00:00:00:00:00:00:01',
+                    dpid_a: {
+                        'dpid': dpid_a,
+                        'enabled': True,
+                        'id': dpid_a,
+                        'interfaces': {
+                            f'{dpid_a}:2': {
+                                'enabled': True,
+                                'lldp': True,
+                                'port_number': 2,
+                                'name': 's1-eth2'
+                            }
+                        }
+                    },
+                    dpid_b: {
+                        'dpid': dpid_b,
                         'enabled': True,
-                        'id': '00:00:00:00:00:00:00:01',
+                        'id': dpid_b,
                         'interfaces': {
-                            '00:00:00:00:00:00:00:01:1': {
+                            f'{dpid_b}:2': {
                                 'enabled': True,
                                 'lldp': True,
-                                'id': '00:00:00:00:00:00:00:00:1',
+                                'port_number': 2,
+                                'name': 's2-eth2'
                             }
                         }
                     }
                 }
             }
         }
-        switches_expected = {'00:00:00:00:00:00:00:01': True}
-        interfaces_expected = {'00:00:00:00:00:00:00:01:1': (True, True)}
+        switches_expected = [dpid_a, dpid_b]
+        interfaces_expected = [f'{dpid_a}:2', f'{dpid_b}:2']
+        links_expected = [link_id]
         mock_storehouse_get_data.return_value = status
         self.napp._load_network_status()
-        self.assertDictEqual(switches_expected, self.napp.switches_state)
-        self.assertDictEqual(interfaces_expected, self.napp.interfaces_state)
-        self.assertDictEqual(status['network_status']['links'],
-                             self.napp.links_state)
-
-    # pylint: disable=too-many-locals
-    def test_restore_network_status(self):
-        """Test restore_network_status."""
-        dpid = '00:00:00:00:00:00:00:01'
-        mock_switch = get_switch_mock(dpid)
-        mock_switch.id = dpid
-        mock_interface = get_interface_mock('s1-eth1', 1, mock_switch)
-        mock_switch.interfaces = {1: mock_interface}
-        self.napp.controller.switches = {dpid: mock_switch}
-        self.napp.switches_state = {dpid: True}
-        self.napp.interfaces_state = {'00:00:00:00:00:00:00:01:1': (True,
-                                                                    True)}
+        self.assertListEqual(switches_expected,
+                             list(self.napp.controller.switches.keys()))
+        interfaces = []
+        for switch in self.napp.controller.switches.values():
+            for iface in switch.interfaces.values():
+                interfaces.append(iface.id)
+        self.assertListEqual(interfaces_expected, interfaces)
+        self.assertListEqual(links_expected, list(self.napp.links.keys()))
 
-        # enable
-        self.napp.restore_network_status(mock_switch)
-        self.assertEqual(mock_switch.enable.call_count, 1)
-        self.assertEqual(mock_interface.enable.call_count, 1)
-        self.assertEqual(mock_interface.lldp, True)
-
-        # disable
-        self.napp.switches_state = {dpid: False}
-        self.napp.interfaces_state = {'00:00:00:00:00:00:00:01:1': (False,
-                                                                    False)}
-        self.napp.restore_network_status(mock_switch)
-        self.assertEqual(mock_switch.disable.call_count, 1)
-        self.assertEqual(mock_interface.disable.call_count, 1)
-        self.assertEqual(mock_interface.lldp, False)
+    @patch('napps.kytos.topology.main.StoreHouse.get_data')
+    @patch('napps.kytos.topology.main.log')
+    def test_load_network_status_fail(self, *args):
+        """Test _load_network_status failure."""
+        (mock_log, mock_storehouse_get_data) = args
+        mock_log.error.return_value = True
+        mock_storehouse_get_data.side_effect = FileNotFoundError('xpto')
+        self.napp._load_network_status()
+        error = 'Fail to load network status from storehouse: xpto'
+        mock_log.error.assert_called_with(error)
 
-    def test_restore_links(self):
-        """Test restore_link."""
-        dpid = '00:00:00:00:00:00:00:01'
-        dpid_b = '00:00:00:00:00:00:00:02'
+    @patch('napps.kytos.topology.main.StoreHouse.get_data')
+    @patch('napps.kytos.topology.main.log')
+    def test_load_network_status_does_nothing(self, *args):
+        """Test _load_network_status doing nothing."""
+        (mock_log, mock_storehouse_get_data) = args
+        mock_log.info.return_value = True
+        mock_storehouse_get_data.return_value = {}
+        self.napp._load_network_status()
+        error = 'There is no status saved to restore.'
+        mock_log.info.assert_called_with(error)
+
+    @patch('napps.kytos.topology.main.StoreHouse.get_data')
+    @patch('napps.kytos.topology.main.Main._load_switch')
+    @patch('napps.kytos.topology.main.log')
+    def test_load_network_status_fail_switch(self, *args):
+        """Test _load_network_status failure in switch."""
+        (mock_log, mock_load_switch, mock_get_data) = args
+        status = {
+            'network_status': {
+                'id': 'network_status',
+                'links': {},
+                'switches': {
+                    '1': {}
+                }
+            }
+        }
+        mock_log.error.return_value = True
+        mock_get_data.return_value = status
+        mock_load_switch.side_effect = Exception('xpto')
+        self.napp._load_network_status()
+        error = 'Error loading switch: xpto'
+        mock_log.error.assert_called_with(error)
+
+    @patch('napps.kytos.topology.main.StoreHouse.get_data')
+    @patch('napps.kytos.topology.main.Main._load_link')
+    @patch('napps.kytos.topology.main.log')
+    def test_load_network_status_fail_link(self, *args):
+        """Test _load_network_status failure in link."""
+        (mock_log, mock_load_link, mock_get_data) = args
+        status = {
+            'network_status': {
+                'id': 'network_status',
+                'switches': {},
+                'links': {
+                    '1': {}
+                }
+            }
+        }
+        mock_log.error.return_value = True
+        mock_get_data.return_value = status
+        mock_load_link.side_effect = Exception('xpto')
+        self.napp._load_network_status()
+        error = 'Error loading link: xpto'
+        mock_log.error.assert_called_with(error)
+
+    @patch('napps.kytos.topology.main.KytosEvent')
+    @patch('kytos.core.buffers.KytosEventBuffer.put')
+    def test_load_switch(self, *args):
+        """Test _load_switch."""
+        (mock_buffers_put, mock_event) = args
+        dpid_a = "00:00:00:00:00:00:00:01"
+        dpid_x = "00:00:00:00:00:00:00:XX"
+        iface_a = f'{dpid_a}:1'
+        switch_attrs = {
+            'dpid': dpid_a,
+            'enabled': True,
+            'id': dpid_a,
+            'interfaces': {
+                iface_a: {
+                    'enabled': True,
+                    'lldp': True,
+                    'id': iface_a,
+                    'switch': dpid_a,
+                    'name': 's2-eth1',
+                    'port_number': 1
+                }
+            }
+        }
+        self.napp._load_switch(dpid_a, switch_attrs)
+
+        self.assertEqual(len(self.napp.controller.switches), 1)
+        self.assertIn(dpid_a, self.napp.controller.switches)
+        self.assertNotIn(dpid_x, self.napp.controller.switches)
+        switch = self.napp.controller.switches[dpid_a]
+
+        self.assertEqual(switch.id, dpid_a)
+        self.assertEqual(switch.dpid, dpid_a)
+        self.assertTrue(switch.is_enabled())
+        self.assertFalse(switch.is_active())
+
+        self.assertEqual(len(switch.interfaces), 1)
+        self.assertIn(1, switch.interfaces)
+        self.assertNotIn(2, switch.interfaces)
+        mock_event.assert_called()
+        mock_buffers_put.assert_called()
+
+        interface = switch.interfaces[1]
+        self.assertEqual(interface.id, iface_a)
+        self.assertEqual(interface.switch.id, dpid_a)
+        self.assertEqual(interface.port_number, 1)
+        self.assertTrue(interface.is_enabled())
+        self.assertFalse(interface.is_active())
+        self.assertTrue(interface.lldp)
+        self.assertTrue(interface.uni)
+        self.assertFalse(interface.nni)
+
+    def test_load_switch_attrs(self):
+        """Test _load_switch."""
+        dpid_b = "00:00:00:00:00:00:00:02"
+        iface_b = f'{dpid_b}:1'
+        switch_attrs = {
+            "active": True,
+            "connection": "127.0.0.1:43230",
+            "data_path": "XX Human readable desc of dp",
+            "dpid": "00:00:00:00:00:00:00:02",
+            "enabled": False,
+            "hardware": "Open vSwitch",
+            "id": "00:00:00:00:00:00:00:02",
+            "interfaces": {
+                "00:00:00:00:00:00:00:02:1": {
+                    "active": True,
+                    "enabled": False,
+                    "id": "00:00:00:00:00:00:00:02:1",
+                    "link": "",
+                    "lldp": False,
+                    "mac": "de:58:c3:30:b7:b7",
+                    "metadata": {},
+                    "name": "s2-eth1",
+                    "nni": False,
+                    "port_number": 1,
+                    "speed": 1250000000,
+                    "switch": "00:00:00:00:00:00:00:02",
+                    "type": "interface",
+                    "uni": True
+                },
+            },
+            "manufacturer": "Nicira, Inc.",
+            "metadata": {},
+            "name": "00:00:00:00:00:00:00:04",
+            "ofp_version": "0x04",
+            "serial": "XX serial number",
+            "software": "2.10.7",
+            "type": "switch"
+        }
+
+        self.napp._load_switch(dpid_b, switch_attrs)
+
+        self.assertEqual(len(self.napp.controller.switches), 1)
+        self.assertIn(dpid_b, self.napp.controller.switches)
+
+        switch = self.napp.controller.switches[dpid_b]
+        self.assertEqual(switch.id, dpid_b)
+        self.assertEqual(switch.dpid, dpid_b)
+        self.assertFalse(switch.is_enabled())
+        self.assertFalse(switch.is_active())
+        self.assertEqual(switch.description['manufacturer'], 'Nicira, Inc.')
+        self.assertEqual(switch.description['hardware'], 'Open vSwitch')
+        self.assertEqual(switch.description['software'], '2.10.7')
+        self.assertEqual(switch.description['serial'], 'XX serial number')
+        self.assertEqual(switch.description['data_path'],
+                         'XX Human readable desc of dp')
+
+        self.assertEqual(len(switch.interfaces), 1)
+        self.assertIn(1, switch.interfaces)
+        self.assertNotIn(2, switch.interfaces)
+
+        interface = switch.interfaces[1]
+        self.assertEqual(interface.id, iface_b)
+        self.assertEqual(interface.switch.id, dpid_b)
+        self.assertEqual(interface.port_number, 1)
+        self.assertFalse(interface.is_enabled())
+        self.assertFalse(interface.is_active())
+        self.assertFalse(interface.lldp)
+        self.assertTrue(interface.uni)
+        self.assertFalse(interface.nni)
+
+    def test_load_link(self):
+        """Test _load_link."""
+        dpid_a = "00:00:00:00:00:00:00:01"
+        dpid_b = "00:00:00:00:00:00:00:02"
+        mock_switch_a = get_switch_mock(dpid_a, 0x04)
+        mock_switch_b = get_switch_mock(dpid_b, 0x04)
+        mock_interface_a = get_interface_mock('s1-eth1', 1, mock_switch_a)
+        mock_interface_a.id = dpid_a + ':1'
+        mock_interface_b = get_interface_mock('s2-eth1', 1, mock_switch_b)
+        mock_interface_b.id = dpid_b + ':1'
+        mock_switch_a.interfaces = {1: mock_interface_a}
+        mock_switch_b.interfaces = {1: mock_interface_b}
+        self.napp.controller.switches[dpid_a] = mock_switch_a
+        self.napp.controller.switches[dpid_b] = mock_switch_b
+        link_attrs = {
+            'enabled': True,
+            'endpoint_a': {
+                'switch': dpid_a,
+                'port_number': 1
+            },
+            'endpoint_b': {
+                'switch': dpid_b,
+                'port_number': 1
+            }
+        }
+
+        self.napp._load_link(link_attrs)
+
+        self.assertEqual(len(self.napp.links), 1)
+        link = list(self.napp.links.values())[0]
+
+        self.assertEqual(link.endpoint_a.id, mock_interface_a.id)
+        self.assertEqual(link.endpoint_b.id, mock_interface_b.id)
+        self.assertTrue(mock_interface_a.nni)
+        self.assertTrue(mock_interface_b.nni)
+        self.assertEqual(mock_interface_a.update_link.call_count, 1)
+        self.assertEqual(mock_interface_b.update_link.call_count, 1)
+
+        # test enable/disable
         link_id = '4d42dc08522'
-        mock_switch_a = get_switch_mock(dpid)
-        mock_switch_b = get_switch_mock(dpid_b)
-        mock_interface_a_1 = get_interface_mock('s1-eth1', 1, mock_switch_a)
-        mock_interface_b_1 = get_interface_mock('s2-eth1', 1, mock_switch_b)
-        mock_link = get_link_mock(mock_interface_a_1, mock_interface_b_1)
+        mock_interface_a = get_interface_mock('s1-eth1', 1, mock_switch_a)
+        mock_interface_b = get_interface_mock('s2-eth1', 1, mock_switch_b)
+        mock_link = get_link_mock(mock_interface_a, mock_interface_b)
         mock_link.id = link_id
-        self.napp.links = {link_id: mock_link}
-        self.napp.links_state = {link_id: {'enabled': True}}
-        # enable link
-        self.napp.restore_network_status(mock_link)
-        self.assertEqual(mock_link.enable.call_count, 1)
+        with patch('napps.kytos.topology.main.Main._get_link_or_create',
+                   return_value=mock_link):
+            # enable link
+            link_attrs['enabled'] = True
+            self.napp.links = {link_id: mock_link}
+            self.napp._load_link(link_attrs)
+            self.assertEqual(mock_link.enable.call_count, 1)
+            # disable link
+            link_attrs['enabled'] = False
+            self.napp.links = {link_id: mock_link}
+            self.napp._load_link(link_attrs)
+            self.assertEqual(mock_link.disable.call_count, 1)
 
-        # disable link
-        self.napp.links_state = {link_id: {"enabled": False}}
-        self.napp._verified_links = []
-        self.napp.restore_network_status(mock_link)
-        self.assertEqual(mock_link.disable.call_count, 1)
-
-    def test_fail_restore_link(self):
-        """Test fail restore_link."""
-        dpid = '00:00:00:00:00:00:00:01'
+    @patch('napps.kytos.topology.main.Main._get_link_or_create')
+    def test_fail_load_link(self, get_link_or_create_mock):
+        """Test fail load_link."""
+        dpid_a = '00:00:00:00:00:00:00:01'
         dpid_b = '00:00:00:00:00:00:00:02'
         link_id = '4d42dc08522'
-        link_id_fail = '4cd52'
-        mock_switch_a = get_switch_mock(dpid)
+        mock_switch_a = get_switch_mock(dpid_a)
         mock_switch_b = get_switch_mock(dpid_b)
         mock_interface_a_1 = get_interface_mock('s1-eth1', 1, mock_switch_a)
         mock_interface_b_1 = get_interface_mock('s2-eth1', 1, mock_switch_b)
         mock_link = get_link_mock(mock_interface_a_1, mock_interface_b_1)
         mock_link.id = link_id
         self.napp.links = {link_id: mock_link}
-        self.napp.links_state = {link_id: {"enabled": True}}
-        with self.assertRaises(RestoreError):
-            self.napp._restore_link(link_id_fail)
-
-        self.napp.links_state = {link_id_fail: {"enabled": True}}
-        with self.assertRaises(RestoreError):
-            self.napp._restore_link(link_id_fail)
-
-    def test_fail_restore_switch(self):
-        """Test fail restore_switch."""
-        dpid = '00:00:00:00:00:00:00:01'
-        dpid_fail = '00:00:00:00:00:00:00:06'
-        mock_switch = get_switch_mock(dpid)
-        mock_switch.id = dpid
-        mock_interface = get_interface_mock('s1-eth1', 1, mock_switch)
-        mock_switch.interfaces = {1: mock_interface}
-        self.napp.controller.switche = {dpid: mock_switch}
-        self.napp.switches_state = {dpid: True}
-        self.napp.interfaces_state = {'00:00:00:00:00:00:00:01:1': (True,
-                                                                    True)}
-
+        get_link_or_create_mock.return_value = mock_link
+
+        link_attrs_fail = {
+            'enabled': True,
+            'endpoint_a': {
+                'switch': dpid_a,
+                'port_number': 999
+            },
+            'endpoint_b': {
+                'switch': dpid_b,
+                'port_number': 999
+            }
+        }
         with self.assertRaises(RestoreError):
-            self.napp._restore_switch(dpid_fail)
-
-        self.napp.switches_state = {dpid_fail: True}
+            self.napp._load_link(link_attrs_fail)
+
+        link_attrs_fail = {
+            'enabled': True,
+            'endpoint_a': {
+                'switch': '00:00:00:00:00:00:00:99',
+                'port_number': 1
+            },
+            'endpoint_b': {
+                'switch': '00:00:00:00:00:00:00:77',
+                'port_number': 1
+            }
+        }
         with self.assertRaises(RestoreError):
-            self.napp._restore_switch(dpid_fail)
+            self.napp._load_link(link_attrs_fail)
 
     @patch('napps.kytos.topology.main.Main.save_status_on_storehouse')
     def test_enable_switch(self, mock_save_status):

From 6006bacd36eef7a1cbccd31433cd14c9b1763210 Mon Sep 17 00:00:00 2001
From: Italo Valcy <italo@ampath.net>
Date: Fri, 27 Aug 2021 11:30:56 -0300
Subject: [PATCH 5/5] adjusting unit tests

---
 tests/unit/test_main.py | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/tests/unit/test_main.py b/tests/unit/test_main.py
index 92639a7..1e372ed 100644
--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -318,7 +318,7 @@ def test_load_network_status_fail_link(self, *args):
         mock_get_data.return_value = status
         mock_load_link.side_effect = Exception('xpto')
         self.napp._load_network_status()
-        error = 'Error loading link: xpto'
+        error = 'Error loading link 1: xpto'
         mock_log.error.assert_called_with(error)
 
     @patch('napps.kytos.topology.main.KytosEvent')
@@ -367,7 +367,6 @@ def test_load_switch(self, *args):
         self.assertEqual(interface.switch.id, dpid_a)
         self.assertEqual(interface.port_number, 1)
         self.assertTrue(interface.is_enabled())
-        self.assertFalse(interface.is_active())
         self.assertTrue(interface.lldp)
         self.assertTrue(interface.uni)
         self.assertFalse(interface.nni)
@@ -437,7 +436,6 @@ def test_load_switch_attrs(self):
         self.assertEqual(interface.switch.id, dpid_b)
         self.assertEqual(interface.port_number, 1)
         self.assertFalse(interface.is_enabled())
-        self.assertFalse(interface.is_active())
         self.assertFalse(interface.lldp)
         self.assertTrue(interface.uni)
         self.assertFalse(interface.nni)
